ngraph/core/src/op/experimental_detectron_prior_grid_generator.cpp

```diff
@@ -0,0 +1,159 @@
+    auto featmap_shape = get_input_partial_shape(featmap_port);
+    auto im_data_shape = get_input_partial_shape(im_data_port);
+
+    if (priors_shape.rank().is_dynamic() || featmap_shape.rank().is_dynamic())
+    {
+        return;
+    }
+
+    NODE_VALIDATION_CHECK(
+        this, priors_shape.rank().get_length() == 2, "Priors rank must be equal to 2.");
+
+    NODE_VALIDATION_CHECK(this,
+                          priors_shape[1].is_static() && priors_shape[1].get_length() == 4u,
+                          "The last dimension of the 'priors' input must be equal to 4. Got: ",
+                          priors_shape[1]);
+
+    NODE_VALIDATION_CHECK(
+        this, featmap_shape.rank().get_length() == 4, "Feature_map rank must be equal to 4.");
+
+    if (im_data_shape.rank().is_dynamic())
+    {
+        return;
+    }
+
+    NODE_VALIDATION_CHECK(
+        this, im_data_shape.rank().get_length() == 4, "Im_data rank must be equal to 4.");
+
+    const auto num_batches_featmap = featmap_shape[0];
+    const auto num_batches_im_data = im_data_shape[0];
+    NODE_VALIDATION_CHECK(this,
+                          num_batches_featmap.same_scheme(num_batches_im_data),
+                          "The first dimension of both 'feature_map' and 'im_data' must match. "
+                          "Feature_map: ",
+                          num_batches_featmap,
+                          "; Im_data: ",
+                          num_batches_im_data);
+}
+
+void op::v6::ExperimentalDetectronPriorGridGenerator::validate_and_infer_types()
+{
+    NGRAPH_OP_SCOPE(v6_ExperimentalDetectronPriorGridGenerator_validate_and_infer_types);
+    auto priors_shape = get_input_partial_shape(priors_port);
+    auto featmap_shape = get_input_partial_shape(featmap_port);
+    auto input_et = get_input_element_type(0);
+
+    validate();
+
+    set_output_size(1);
+    PartialShape out_shape = {Dimension::dynamic(), Dimension::dynamic(), Dimension::dynamic(), 4};
+    if (m_attrs.flatten)
+    {
+        out_shape = PartialShape{Dimension::dynamic(), 4};
+    }
+
+    if (priors_shape.rank().is_dynamic() || featmap_shape.rank().is_dynamic())
+    {
+        set_output_type(0, input_et, out_shape);
+        return;
+    }
+
+    auto num_priors = priors_shape[0];
+    auto featmap_height = featmap_shape[2];
+    auto featmap_width = featmap_shape[3];
+
+    if (num_priors.is_dynamic() || featmap_height.is_dynamic() || featmap_width.is_dynamic())
+    {
+        if (!m_attrs.flatten)
+        {
+            out_shape[0] = featmap_height;
+            out_shape[1] = featmap_width;
+            out_shape[2] = num_priors;
+        }
+
+        set_output_type(0, input_et, out_shape);
+        return;
+    }
+
+    const size_t priors_num = static_cast<size_t>(num_priors.get_length());
+    const size_t grid_height = static_cast<size_t>(featmap_height.get_length());
+    const size_t grid_width = static_cast<size_t>(featmap_width.get_length());
```

Comment:
I think that this check is not needed. Even if some of the dimensions are dynamic we can perform manipulations with them as usual, so the expression: `out_shape = Shape{grid_height * grid_width * priors_num, 4};` is correct even if the `grid_height`, `grid_width` and `priors_num` are dynamic. So you can use values `num_priors`, `featmap_height` and `featmap_width` in lines 152 and 156.

Reply:
Please, modify the code and add some tests that the shape interval for dynamic dimensions is calculated correctly.

Reply:
Done.